ratelimits/utilities.go

```diff
@@ -24,49 +24,75 @@ func joinWithColon(args ...string) string {
-					// If the public suffix is the domain itself, that's fine.
-					// Include the original name in the result.
-					covers = append(covers, ident.Value)
-					continue
-				} else {
-					return nil, err
-				}
-			}
-			covers = append(covers, domain)
-		case identifier.TypeIP:
-			ip, err := netip.ParseAddr(ident.Value)
-			if err != nil {
-				return nil, err
-			}
-			prefix, err := coveringPrefix(ip)
-			if err != nil {
-				return nil, err
-			}
-			covers = append(covers, prefix.String())
+		cover, err := coveringIdentifier(CertificatesPerDomain, ident)
+		if err != nil {
+			return nil, err
 		}
+		covers = append(covers, cover)
 	}
 	return core.UniqueLowerNames(covers), nil
 }
 
-// coveringPrefix transforms a netip.Addr into its "covering" prefix, for the
-// CertificatesPerDomain limit. For IPv4, this is the IP address (/32). For
-// IPv6, this is the /64 that contains the address.
-func coveringPrefix(addr netip.Addr) (netip.Prefix, error) {
-	var bits int
-	if addr.Is4() {
-		bits = 32
-	} else {
-		bits = 64
+// coveringIdentifier transforms a single ACMEIdentifier into its "covering"
+// identifier, for the CertificatesPerDomain, CertificatesPerDomainPerAccount,
+// and NewRegistrationsPerIPv6Range limits. For DNS identifiers, this is the
+// eTLD+1; exact public suffix matches are included. For IP address identifiers,
+// this is the address (/32) for IPv4, or the /64 prefix for IPv6, in CIDR
+// notation.
+func coveringIdentifier(limit Name, ident identifier.ACMEIdentifier) (string, error) {
+	switch ident.Type {
+	case identifier.TypeDNS:
+		domain, err := publicsuffix.Domain(ident.Value)
+		if err != nil {
+			if err.Error() == fmt.Sprintf("%s is a suffix", ident.Value) {
+				// If the public suffix is the domain itself, that's fine.
+				// Include the original name in the result.
+				return ident.Value, nil
+			}
+			return "", err
+		}
+		return domain, nil
+	case identifier.TypeIP:
+		ip, err := netip.ParseAddr(ident.Value)
+		if err != nil {
+			return "", err
+		}
+		prefix, err := coveringPrefix(limit, ip)
+		if err != nil {
+			return "", err
+		}
+		return prefix.String(), nil
 	}
-	prefix, err := addr.Prefix(bits)
-	if err != nil {
-		// This should be impossible because bits is hardcoded.
-		return netip.Prefix{}, err
+	return "", fmt.Errorf("unsupported identifier type: %s", ident.Type)
+}
+
+// coveringPrefix transforms a netip.Addr into its "covering" prefix, for the
+// CertificatesPerDomain, CertificatesPerDomainPerAccount, and
+// NewRegistrationsPerIPv6Range limits.
+func coveringPrefix(limit Name, addr netip.Addr) (netip.Prefix, error) {
+	switch limit {
+	case CertificatesPerDomain, CertificatesPerDomainPerAccount:
+		var bits int
+		if addr.Is4() {
+			bits = 32
+		} else {
+			bits = 64
+		}
+		prefix, err := addr.Prefix(bits)
+		if err != nil {
+			return netip.Prefix{}, fmt.Errorf("building covering prefix for %s: %w", addr, err)
+		}
+		return prefix, nil
+
+	case NewRegistrationsPerIPv6Range:
```

Comment:
I feel a little uneasy about moving this check all the way into this function; it's sort of an unexpected place for it to live when thinking about the flow.

Reply:
It also means we can stop doing this at bucketKey creation time. https://github.com/letsencrypt/boulder/pull/8279/files#diff-249a890ad22c401d8fdd218f86feaadaf7e5d07cdacfea8a8c701c08c54acb8eL25-L35